File: Other/LongestSubstringWithoutRepeatingCharacters.py
from sets import Set

def longest_substring(s):
	if not s:
		return None

	start = 0
	end = 0

	longest = ''
	char_set = Set()

	while end < len(s):
		end += 1
		cur_char = s[end - 1]

		#print "cur_char: " + str(cur_char)
		if cur_char not in char_set:
			char_set.add(cur_char)
			#print char_set

			if end - start > len(longest):
				longest = s[start: end]
			
			#print "longest: " + str(longest)
			#print '------------------------'
			continue

		while start < end - 1:
			if s[start] != cur_char:
				char_set.remove(s[start])
				#print "*****Removing"
				#print s[start]
				#print char_set
				start += 1
			else:
				start += 1
				break
		#print "longest: " + str(longest)
		#print '------------------------'
	return longest;

print longest_substring("bacadbef")
print longest_substring("aabbaa")
print longest_substring("aabbbaa")
print longest_substring("axcaxc")
print longest_substring("abkbcd")
print longest_substring("aaaaaa")

File: Other/LongestSubstringWithoutRepeatingCharacters.java
/*
	Given a string, find the longest substring without repeating characters.
*/
import java.util.HashSet;
import java.util.Set;

class LongestSubstringWithoutRepeatingCharacters {

	public static String find(String str) {
		if (str.length() <= 1) {
			return str;
		}

		int start = 0;
		String result = "";
		Set<Character> char_set = new HashSet<>();

		for (int end = 0; end < str.length(); end++) {
			
			if (!char_set.contains(str.charAt(end))) {
				char_set.add(str.charAt(end));
				if ((end + 1 - start) > result.length()) {
					result = str.substring(start, end + 1);
				}
				continue;
			}

			while ( start < end) {
				if (str.charAt(start) != str.charAt(end)) {
					char_set.remove(str.charAt(start));
					start++;
				} else {
					start++;
					break;
				}
			}

		}

		return result;
	}

	public static void main(String[] args) {

		System.out.println(find("aabbaa").equals("ab") + " - " + find("aabbaa"));
		System.out.println(find("aabbbaa").equals("ab") + " - " + find("aabbbaa"));
		System.out.println(find("axcaxc").equals("axc") + " - " + find("axcaxc"));
		System.out.println(find("abcadbef").equals("cadbef") + " - " + find("abcadbef"));
		System.out.println(find("abkbcd").equals("kbcd") + " - " + find("abkbcd"));
		System.out.println(find("aaaaaa").equals("a") + " - " + find("aaaaaa"));
		System.out.println(find("pwwkew").equals("wke") + " - " + find("pwwkew"));
		System.out.println(find("abcabcbb").equals("abc") + " - " + find("abcabcbb"));

	}
}